### ISSUE TITLE
is --net=host  really needed?

### ISSUE THREAD

**Issue by paspo** [rustdesk/rustdesk-server#52]
without --net=host, our P2P direct link (via hole punch) can not work, because hbbs can only get virtual ip of docker rathter than real ip of the connection.

_Originally posted by @rustdesk in https://github.com/rustdesk/rustdesk-server/issues/36#issuecomment-1172285170_

**Comment by paspo** [comment]
I just happened to have a couple of brand new Windows 10 pro VMs laying around, so I did a quick test: I put each VM in a separate network that can only go to internet, but can't reach the other network directly.

Everything works, as expected.

With hole-punching you don't need to rely on the IP you get for the connection, because it' very likely you're dealing with some NAT (at least 1, but you can have CGN also). That's the whole point of hole-punching: reaching devices behind NATs. The "internal docker network" is essentially a private network that exposes your container's ports and is able to NAT those connections.

Please, @rustdesk, help me understand.

Logs here, for reference:

```
rustdesk    | /package/admin/s6-overlay/libexec/preinit: notice: /var/run is not a symlink to /run, fixing it
rustdesk    | s6-rc: info: service hbbs: starting
rustdesk    | s6-rc: info: service hbbr: starting
rustdesk    | s6-rc: info: service s6rc-oneshot-runner: starting
rustdesk    | [2022-07-01 15:15:40.413957 +00:00] INFO [src/relay_server.rs:60] #blacklist(blacklist.txt): 0
rustdesk    | [2022-07-01 15:15:40.413978 +00:00] INFO [src/relay_server.rs:75] #blocklist(blocklist.txt): 0
rustdesk    | [2022-07-01 15:15:40.413982 +00:00] INFO [src/relay_server.rs:81] Listening on tcp 0.0.0.0:21117
rustdesk    | [2022-07-01 15:15:40.413985 +00:00] INFO [src/relay_server.rs:83] Listening on websocket 0.0.0.0:21119
rustdesk    | [2022-07-01 15:15:40.413988 +00:00] INFO [src/relay_server.rs:85] Start
rustdesk    | [2022-07-01 15:15:40.414021 +00:00] INFO [src/relay_server.rs:104] DOWNGRADE_THRESHOLD: 0.66
rustdesk    | [2022-07-01 15:15:40.414026 +00:00] INFO [src/relay_server.rs:113] DOWNGRADE_START_CHECK: 1800s
rustdesk    | [2022-07-01 15:15:40.414035 +00:00] INFO [src/relay_server.rs:122] LIMIT_SPEED: 4Mb/s
rustdesk    | [2022-07-01 15:15:40.414038 +00:00] INFO [src/relay_server.rs:132] TOTAL_BANDWIDTH: 1024Mb/s
rustdesk    | [2022-07-01 15:15:40.414041 +00:00] INFO [src/relay_server.rs:146] SINGLE_BANDWIDTH: 16Mb/s
rustdesk    | [2022-07-01 15:15:40.416447 +00:00] INFO [src/peer.rs:82] DB_URL=./db_v2.sqlite3
rustdesk    | [2022-07-01 15:15:40.417705 +00:00] INFO [src/rendezvous_server.rs:94] serial=0
rustdesk    | [2022-07-01 15:15:40.417714 +00:00] INFO [src/common.rs:41] rendezvous-servers=[]
rustdesk    | [2022-07-01 15:15:40.417718 +00:00] INFO [src/rendezvous_server.rs:96] Listening on tcp/udp 0.0.0.0:21116
rustdesk    | [2022-07-01 15:15:40.417720 +00:00] INFO [src/rendezvous_server.rs:97] Listening on tcp 0.0.0.0:21115, extra port for NAT test
rustdesk    | [2022-07-01 15:15:40.417723 +00:00] INFO [src/rendezvous_server.rs:98] Listening on websocket 0.0.0.0:21118
rustdesk    | [2022-07-01 15:15:40.417744 +00:00] INFO [libs/hbb_common/src/udp.rs:33] Receive buf size of udp 0.0.0.0:21116: Ok(212992)
rustdesk    | [2022-07-01 15:15:40.417802 +00:00] INFO [src/common.rs:108] Private key comes from id_ed25519
rustdesk    | s6-rc: info: service s6rc-oneshot-runner successfully started
rustdesk    | s6-rc: info: service fix-attrs: starting
rustdesk    | s6-rc: info: service hbbs successfully started
rustdesk    | s6-rc: info: service hbbr successfully started
rustdesk    | s6-rc: info: service fix-attrs successfully started
rustdesk    | s6-rc: info: service legacy-cont-init: starting
rustdesk    | s6-rc: info: service legacy-cont-init successfully started
rustdesk    | s6-rc: info: service legacy-services: starting
rustdesk    | s6-rc: info: service legacy-services successfully started
rustdesk    | [2022-07-01 15:15:40.522944 +00:00] INFO [src/common.rs:41] relay-servers=["you.dont.need.to.know.my.relay:21117"]
rustdesk    | [2022-07-01 15:15:40.523193 +00:00] INFO [src/rendezvous_server.rs:135] ALWAYS_USE_RELAY=N
rustdesk    | [2022-07-01 15:15:40.523215 +00:00] INFO [src/rendezvous_server.rs:155] Start
rustdesk    | [2022-07-01 15:15:40.546343 +00:00] INFO [src/peer.rs:128] pk updated instead of insert
rustdesk    | [2022-07-01 15:16:52.198462 +00:00] INFO [src/peer.rs:100] update_pk 166581618 [my-ip-address]59611 [52, 101, 48, 99, 99, 100, 55, 102, 45, 102, 55, 51, 51, 45, 52, 50, 97, 97, 45, 98, 99, 101, 101, 45, 54, 49, 50, 57, 53, 55, 99, 102, 49, 54, 55, 97] [58, 62, 183, 131, 57, 213, 77, 145, 191, 25, 92, 166, 248, 22, 104, 37, 230, 97, 67, 126, 194, 213, 49, 57, 110, 116, 223, 178, 192, 235, 3, 38]
rustdesk    | [2022-07-01 15:16:58.499927 +00:00] INFO [src/peer.rs:100] update_pk 164411550 [my-ip-address]:55487 [99, 51, 51, 98, 49, 49, 56, 48, 45, 48, 99, 51, 54, 45, 52, 51, 57, 98, 45, 98, 101, 100, 102, 45, 54, 50, 99, 56, 52, 101, 49, 98, 99, 99, 56, 49] [221, 109, 14, 56, 62, 206, 248, 142, 121, 97, 253, 4, 206, 113, 84, 135, 185, 174, 238, 202, 163, 129, 66, 201, 70, 121, 92, 0, 247, 170, 154, 15]
rustdesk    | [2022-07-01 15:18:27.593568 +00:00] INFO [src/relay_server.rs:446] New relay request b58f8b40-a8f7-4fc9-b042-9976b65d0690 from [my-ip-address]:46077
rustdesk    | [2022-07-01 15:18:29.841083 +00:00] INFO [src/relay_server.rs:430] Relayrequest b58f8b40-a8f7-4fc9-b042-9976b65d0690 from [my-ip-address]:4559 got paired
rustdesk    | [2022-07-01 15:18:29.841123 +00:00] INFO [src/relay_server.rs:436] Both are raw
```

**Comment by rustdesk** [comment]
It is connection but not direct connection, it is relayed by hbbr.

**Comment by paspo** [comment]
Totally missed the `direct` part on your comments, sorry. And after re-reading my comments, I realized I made it even worse...

The important difference (for us) between `--net=host` and `--net=bridge` in docker is that in the first case the container is directly accessing the networking of the host (as if it is a normal running application) and in the last the container network layer is "confined" aka an internal bridge interface is created and all traffic is natted. Bridge mode is great for container isolation and to avoid port duplication (you can have 10 containers listening on port 3000 on the internal bridge and on different ports on the host).
Unfortunately, bridge mode comes with a NAT and AFAIK there's no reliable direct way to know the real IP behind a connection from the perspective of a software running inside the container.

That said, I tried:

- customizing networks disabling ip_masquerade -> no luck
- adding capabilities (NET_ADMIN, NET_RAW and even ALL) -> no luck 
- defining custom networks and playing with iptables packet mangling -> kinda works, really clunky and unreliable

For the moment, there's no better way to achieve direct connections than `--net=host`. 

IMHO, direct connection is rarely needed, BTW: 

- if you're doing help-desk remote assistance, you're not on the same net anyway
- same if you're accessing your personal pc or your friends pc from remote
- if you're doing help-desk remote assistance for your company it's very likely that you're on the same network (and want direct connection) and you're using direct IP connection

The biggest drawback is that this issue is not solvable on some containerization environments (k8s for example)

To correctly implement host networking with docker-compose, you do this:

```yaml
version: '3'

services:
  rustdesk-server:
    container_name: rustdesk-server
    network_mode: host
    image: rustdesk/rustdesk-server-s6:latest
    environment:
      - "RELAY=rustdesk.example.com:21117"
      - "ENCRYPTED_ONLY=1"
    volumes:
      - ./data:/data
    restart: unless-stopped
```

**Comment by denis-domanskii** [comment]
Hi @paspo! Could you please clarify, if we need `--net=host` only to get the info about real IP, then why we can't set it manually using an env variable?

My network architecture is:
container -> home server -> router -> internet

Container IP: 10.0.0.1 (assigned by the docker)
Home server IP: 192.168.1.2 (assigned by the router)
Router IP: 86.114.52.19 (external IP, accessible from the internet)

So, even if I use `--net=host` here, my home server IP will be still internal, behind the router NAT. What the difference between router only NAT and router + docker NATs? 

From the other hand, I can setup port forwarding on my router, and expose rustdesk server port. It will be something like:

rustdesk server (docker compose):
```
ports:
      - 21115:21115
      - 21116:21116
      - 21116:21116/udp
      - 21118:21118
```
which make rustdesk accessible by home server address 192.168.1.2:21115 (21116, 21118)

and then I can setup my router to forward all income connection:
21115 -> 192.168.1.2:21115
21116 -> 192.168.1.2:21116
21118 -> 192.168.1.2:21118

which make rustdesk accessible from the internet. And if rustdesk need to provide it's address to clients, I can set it manually to my external IP: 86.114.52.19. And all clients from the internet will be able to access it. 

So, why I need `--net=host`? Maybe I didn't understand punch hole concept?

**Comment by paspo** [comment]
Hi @denis-domanskii.

TLDR:
- docker bridge -> always a relay connection
- hbbs behind a router -> always a relay connection
- directly exposed `hbbs` -> direct connection if possible, otherwise relay



It all comes down to identify the source of the connection. If the `hbbs` service is "directly exposed", it can know the source address of the connection; on the other hand, if there's a "network" in between (a NAT), from the point of view of the `hbbs` service, the connection is always coming from the same IP.
This is true for docker in bridge mode (the mode in which you have to expose the ports) AND for a host (or a VM) running inside your network, behind your firewall; in fact, you are port forwarding in your firewall, exactly like you do for a docker container. It's a NAT either way.
So, every time you have an address translation, you're "masking" the real originating IP address of the connection.

The consequence of this is that `hbbs` will decide to handle this connection through a relay, so it'll force a session to pass through an `hbbr` instance.

In your case, you have at least 2 NAT: the home router and the docker in bridge mode running on tour home server. It's better to have a VM in a datacenter somewhere.
(I'm assuming you want to use rustdesk to connect to hosts outside your home network)
